ZukoAzula2016/src/org/team1540/drivers/kauailabs/AHRSRegister.java

```diff
@@ -0,0 +1,197 @@
+        if (registerID < 0 || registerID >= 128 || rt == null || mode == null) {
+            throw new IllegalArgumentException();
+        }
+        if (extra != 0 && rt != RegisterType.FLAG16 && rt != RegisterType.U16_MUL) {
+            throw new IllegalArgumentException(); // must not pass extra
+        }
+        this.extra = extra;
+        this.registerID = (byte) registerID;
+        this.type = rt;
+        this.mode = mode;
+    }
+
+    private AHRSRegister(int registerID, RegisterType rt, WriteMode mode) {
+        this(registerID, 0, rt, mode);
+        if (rt == RegisterType.FLAG16) {
+            throw new IllegalArgumentException(); // must use previous
+                                                  // constructor
+        }
+    }
+
+    private AHRSRegister(int registerID, RegisterType rt) {
+        this(registerID, rt, WriteMode.READ_ONLY);
+    }
+
+    private AHRSRegister(int registerID) {
+        this(registerID, RegisterType.U8, WriteMode.READ_ONLY);
+    }
+
+    public byte decodeByte(RegisterIO io) {
+        if (type != RegisterType.U8) {
+            throw new RuntimeException("Wrong type!");
+        }
+        return io.getRaw()[registerID - io.getOrigin()];
+    }
+
+    public short decodeShort(RegisterIO io) {
+        if (type != RegisterType.U16) {
+            throw new RuntimeException("Wrong type!");
+        }
+        return AHRSProtocol.decodeBinaryInt16(io.getRaw(), registerID - io.getOrigin());
+    }
+
+    public boolean decodeFlag(RegisterIO io) {
+        if (type != RegisterType.FLAG16) {
+            throw new RuntimeException("Wrong type!");
+        }
+        return (AHRSProtocol.decodeBinaryInt16(io.getRaw(), registerID - io.getOrigin()) & (int) extra) != 0;
+    }
+
+    public int decodeInt(RegisterIO io) {
+        if (type != RegisterType.U32) {
+            throw new RuntimeException("Wrong type!");
+        }
+        return AHRSProtocol.decodeBinaryInt32(io.getRaw(), registerID - io.getOrigin());
+    }
+
+    public int decodeAnyInt(RegisterIO io) {
+        switch (type) {
+        case U8:
+            return decodeByte(io);
+        case U16:
+            return decodeShort(io);
+        case U32:
+            return decodeInt(io);
+        case FLAG16:
+            return decodeFlag(io) ? 1 : 0;
+        default:
+            throw new RuntimeException("Wrong type!");
+        }
+    }
+
+    public int decodeAnyUint(RegisterIO io) {
+        switch (type) {
+        case U8:
+            return decodeByte(io) & 0xFF;
+        case U16:
+            return decodeShort(io) & 0xFFFF;
+        case U32:
+            return decodeInt(io);
+        case FLAG16:
+            return decodeFlag(io) ? 1 : 0;
+        default:
+            throw new RuntimeException("Wrong type!");
+        }
+    }
+
+    public float decodeAnyFloat(RegisterIO io) {
+        switch (type) {
+        case U8:
+            return decodeByte(io);
```

Comment:
These will all be signed, which is fine, but it should probably be documented because some people might assume that it should be unsigned.